[IMP] web_editor: Image shapes for themes & website configurator

xO-Tx opened the pull request:
The goal of this PR is to add a "Python version" of the
shape on image feature using a controller.

Since the whole logic to apply shapes on image is JS based, we need
to use this URL (just like background shapes) when we want to add
a shape by default on images in themes.

When the configurator replaces a snippet image (with default shape),
the shape option should still be applied on the new one.

task-2593454

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/73938).

qsm-odoo commented on addons/web_editor/controllers/main.py:
```diff
@@ -521,64 +523,97 @@ def modify_image(self, attachment, res_model=None, res_id=None, name=None, data=
+    @http.route(['/web_editor/image_shape/<module>/<path:filename>'], type='http', auth="public", website=True)
+    def image_shape(self, module, filename, img_key, **kwargs):
+        svg = self._get_shape_svg(module, 'static', 'image_shapes', filename)
+        if img_key:
```
Kinda strange to have a required attribute as additional parameter, maybe change the URL form to:

`/web_editor/image_shape/<the image key here>/<module>/<path:filename>`

qsm-odoo replied:
@xO-Tx The point of the URL change, was to be able to remove that condition, you now always have an `img_key`.

qsm-odoo reviewed (changes requested):
Seems to be a good work, try to finish it even faster, I'd like to merge this today :+1: You'll have to adapt some stuff according to my only comment. Also in the second commit you say:

> - Update 'data-mimetype' on image to SVG since loadImageInfo()
> will set the original attachment mimetype on it.

-> seems useless to mention in the commit message

> - Prevent saving image as base64 when it's saved without
any changes being made.

I don't see what part of the code is doing that? And should not it be in a separate fix commit then?

Squash both commits, they are very related IMO.

xO-Tx commented:
> > * Prevent saving image as base64 when it's saved without
> >   any changes being made.
> 
> I don't see what part of the code is doing that? And should not it be in a separate fix commit then?

This happens when image has a shape option + image src is not a data URL (e.g. image with a saved shape option / image with shape added by theme using the Python URL).

I fixed that by adding a condition to prevent `_applyOptions()` from reapplying shapes (which converts image to base64) when image URL returns **image + a shape applied on it**.

Moved now to a separate commit since it's a general issue.

qsm-odoo commented on addons/web_editor/controllers/main.py:
```diff
@@ -540,14 +541,27 @@ def _update_svg_colors(self, options, svg):
-                if not re.match(r'(?i)^#[0-9A-F]{6,8}$|^rgba?\(\d{1,3},\d{1,3},\d{1,3}(?:,[0-9.]{1,4})?\)$', value.replace(' ', '')):
-                    raise werkzeug.exceptions.BadRequest()
-                user_colors.append([tools.html_escape(value), colorMatch.group(1)])
+                if not re.match(r'(?i)^%s$|^%s$' % (regex_hex, regex_rgba), css_color_value.replace(' ', '')):
+                    if re.match('^o-color-([1-5])$', css_color_value):
+                        if not bundle_css:
+                            bundle = 'web.assets_frontend'
+                            files, _ = request.env["ir.qweb"]._get_asset_content(bundle, options=request.context)
+                            asset = AssetsBundle(bundle, files)
+                            bundle_css = asset.css().index_content
+                        color_search = re.search(r'(?i)--%s:\s+(%s|%s)' % (css_color_value, regex_hex, regex_rgba), bundle_css)
+                        if color_search:
+                            css_color_value = color_search.group(1)
+                    else:
+                        raise werkzeug.exceptions.BadRequest()
+                user_colors.append([tools.html_escape(css_color_value), colorMatch.group(1)])
```
Seems good :+1: but maybe a small problem: when you search for a color in the bundle and that you don't find it (`if color_search` is falsy), you should return a bad request too I think?

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -4797,6 +4797,8 @@ registry.ImageOptimize = ImageHandlerOption.extend({
             // Image data-mimetype should be changed to SVG since loadImageInfo()
             // will set the original attachment mimetype on it.
             img.dataset.mimetype = 'image/svg+xml';
+            img.dataset.shapeColors = img.dataset.shapeColors.split(';')
+                .map(color => this._getCSSColorValue(color)).join(';');
```
Not sure this is perfect: don't you have the risk of ending up saving an image where the URL is still something with `?c1=o-color-3` but where you replaced the data-shape-colors with the actual CSS color of o-color-3?

= Can't you just convert with `_getCSSColorValue` when you **use** that data-shape-colors value?

qsm-odoo reviewed (approved):
Thanks for the work, in the end I squashed the 3 commits and added the final xpath example in the commit message (the 3 commits made the review easier but they were really too related to be merged individually).

@robodoo r+

robodoo commented:
@qsm-odoo, you may want to rebuild or fix this PR as it has failed CI.

qsm-odoo commented:
@xO-Tx A bit better this way: https://github.com/odoo/odoo/compare/17f7e8a30a47acd1aff82c77361b714a1dd76df6..4cad1ef8380a3fac7883e30ac863514ccd879163 :+1:

JKE-be commented:
@robodoo override=ci/security

Code moved
Already quickly checked by odo  https://github.com/odoo/odoo/pull/56479
Always limited to static files, so should be not risky
And now limited to svg

qsm-odoo commented:
@robodoo r+
